tests/safety/test_ford.py

```diff
@@ -0,0 +1,129 @@
+  STANDSTILL_THRESHOLD = 1
+  RELAY_MALFUNCTION_ADDR = MSG_IPMA_DATA
+  RELAY_MALFUNCTION_BUS = 0
+
+  def setUp(self):
+    self.packer = None
+    raise unittest.SkipTest
+
+  # Driver brake pedal
+  def _user_brake_msg(self, brake):
+    # brake pedal and cruise state share same message, so we have to send
+    # the other signal too
+    enable = self.safety.get_controls_allowed()
+    values = {
+      "BpedDrvAppl_D_Actl": 2 if brake else 1,
+      "CcStat_D_Actl": 5 if enable else 0,
+    }
+    return self.packer.make_can_msg_panda("EngBrakeData", 0, values)
+
+  # Vehicle speed
+  def _speed_msg(self, speed):
+    values = {"Veh_V_ActlEng": speed}
+    return self.packer.make_can_msg_panda("EngVehicleSpThrottle2", 0, values)
+
+  # Drive throttle input
+  def _user_gas_msg(self, gas):
+    values = {"ApedPos_Pc_ActlArb": gas}
+    return self.packer.make_can_msg_panda("EngVehicleSpThrottle", 0, values)
+
+  # Cruise status
+  def _pcm_status_msg(self, enable):
+    # brake pedal and cruise state share same message, so we have to send
+    # the other signal too
+    brake = self.safety.get_brake_pressed_prev()
+    values = {
+      "BpedDrvAppl_D_Actl": 2 if brake else 1,
+      "CcStat_D_Actl": 5 if enable else 0,
+    }
+    return self.packer.make_can_msg_panda("EngBrakeData", 0, values)
+
+  # LKAS command
+  def _lkas_command_msg(self, action):
+    values = {
+      "LkaActvStats_D2_Req": action,
+    }
+    return self.packer.make_can_msg_panda("Lane_Assist_Data1", 0, values)
+
+  # TJA command
+  def _tja_command_msg(self, enabled):
+    values = {
+      "LatCtl_D_Rq": 1 if enabled else 0,
+    }
+    return self.packer.make_can_msg_panda("LateralMotionControl", 0, values)
+
+  # Cruise control buttons
+  def _acc_button_msg(self, cancel=0, resume=0, _set=0):
+    values = {
+      "CcAslButtnCnclPress": cancel,
+      "CcAsllButtnResPress": resume,
+      "CcAslButtnSetPress": _set,
+    }
+    return self.packer.make_can_msg_panda("Steering_Data_FD1", 0, values)
+
+
+class TestFordSteeringSafety(TestFordSafety):
+  def setUp(self):
+    self.packer = CANPackerPanda("ford_lincoln_base_pt")
+    self.safety = libpandasafety_py.libpandasafety
+    self.safety.set_safety_hooks(Panda.SAFETY_FORD, 0)
+    self.safety.init_tests()
+
+  def test_steer_when_enabled(self):
+    self.safety.set_controls_allowed(1)
+    self.assertTrue(self._tx(self._tja_command_msg(1)))
+    self.assertTrue(self.safety.get_controls_allowed())
+
+  def test_steer_when_disabled(self):
+    self.safety.set_controls_allowed(0)
+    self.assertFalse(self._tx(self._tja_command_msg(0)))
+    self.assertFalse(self.safety.get_controls_allowed())
+
+  def test_prevent_lkas_action(self):
+    self.safety.set_controls_allowed(1)
+    self.assertFalse(self._tx(self._lkas_command_msg(1)))
+
+    self.safety.set_controls_allowed(0)
+    self.assertFalse(self._tx(self._lkas_command_msg(1)))
+
+
+class TestFordStockSafety(TestFordSafety):
```

Comment:
why are there three different classes? only one is necessary here

Reply:
I was just following the Tesla and VW structure where they look to be split up by functionality

Reply:
The other Tesla classes are for different configurations, e.g. stock long and openpilot long. Not sure why VW has that extra class.